### PR TITLE
feat!: split network step

### LINKED ISSUE #671: Split common stage into two separate stages implementing common interface
### TL;DR

Size: L

### Terraform Resources

_No response_

### Detailed design

_No response_

### Additional information

_No response_

### LINKED ISSUE #672: Test pipeline refactor after networking stage decomposition
### TL;DR

Size: S

### Terraform Resources

_No response_

### Detailed design

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by mauro-cit**
Fixes #672 
Fixes #671 

Some notes about the changes done by this PR:
- As we have separated the step 3-networks, the guide for doing that was through the **enable_hub_and_spoke** variable.
- Basically the resources that required the variable enable_hub_and_spoke as true we put on **3-networks-hub-and-spoke**, the resources that doesn't we put on **3-networks-dual-svpc**. We also moved the resources that were depending on these ones.
- We also removed the variable **enable_hub_and_spoke_transitivity** from the Dual Shared VPC mode since it was only used when the **enable_hub_and_spoke** variable was true
- A script called **restore_tf_files.sh** was also created to undo the actions that the "disable_tf_files.sh" was doing to the *.tfvars and backend files

### CODE REVIEW

File: 3-networks-hub-and-spoke/shared.auto.example.tfvars [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// List of IPv4 address of target name servers for the forwarding zone configuration.
+// See https://cloud.google.com/dns/docs/overview#dns-forwarding-zones
+target_name_server_addresses = ["192.168.0.1", "192.168.0.2"]
```
Comment (daniel-cit) [inline 3-networks-hub-and-spoke/shared.auto.example.tfvars:19]:
```suggestion
target_name_server_addresses = ["192.168.0.1", "192.168.0.2"]

//enable_hub_and_spoke_transitivity = true
```

missing the enable_hub_and_spoke_transitivity  variable in the right file example file

File: 3-networks-hub-and-spoke/README.md [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+Please refer to [troubleshooting](../docs/TROUBLESHOOTING.md) if you run into issues during this step.
+
+## Usage
+
+**Note:** If you are using MacOS, replace `cp -RT` with `cp -R` in the relevant
+commands. The `-T` flag is needed for Linux, but causes problems for MacOS.
+
+### Networking Architecture
+
+This step makes use of the **Dual Shared VPC** architecture mode, and more details can be found described at the **Networking** section of the [Google cloud security foundations guide](https://cloud.google.com/architecture/security-foundations/networking). To see the version that makes use the Hub and Spoke mode, check the step [3-networks-hub-and-spoke](../3-networks-hub-and-spoke).
```
Comment (daniel-cit) [inline 3-networks-hub-and-spoke/README.md:90]:
```suggestion
This step uses the **Hub and Spoke** architecture  mode.
More details can be found at the **Networking** section of the [Google cloud security foundations guide](https://cloud.google.com/architecture/security-foundations/networking#hub-and-spoke).
To enabled **Hub and Spoke** [transitivity](https://cloud.google.com/architecture/security-foundations/networking#hub-and-spoke_transitivity) set the variable `enable_hub_and_spoke_transitivity ` to `true`.
To see the version that makes use of the **Dual Shared VPC** architecture mode check the step [3-networks-dual-svpc](../3-networks-dual-svpc).
```

**Review by daniel-cit — CHANGES_REQUESTED** [review]
The Hub & Spoke Transitivity module

https://github.com/mauro-cit/terraform-example-foundation/tree/feature/split-network-step/3-networks-dual-svpc/modules/transitivity

should be removed from the dual svpc version

there are also some fixes on the readme and `shared.auto.example.tfvars` files in the Hub and Spoke version

File: 3-networks-hub-and-spoke/README.md [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+For an overview of the architecture and the parts, see the
+[terraform-example-foundation README](https://github.com/terraform-google-modules/terraform-example-foundation).
+
+## Purpose
+
+The purpose of this step is to:
+
+- Set up the global [DNS Hub](https://cloud.google.com/blog/products/networking/cloud-forwarding-peering-and-zones).
+- Set up base and restricted shared VPCs with default DNS, NAT (optional), Private Service networking, VPC service controls, on-premises Dedicated or Partner Interconnect, and baseline firewall rules for each environment.
```
Comment (bharathkkb) [inline 3-networks-hub-and-spoke/README.md:62]: Lets update this to reflect hub and spoke model i.e base and restricted hubs and corresponding spokes.

File: 3-networks-hub-and-spoke/envs/shared/README.md [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# 3-networks/shared
+# 3-networks-hub-and-spoke/shared
```
Comment (bharathkkb) [inline 3-networks-hub-and-spoke/envs/shared/README.md:1]: Let's also call out below that hubs are setup here.

**Review by bharathkkb — COMMENTED** [review]
Thanks for the PR @mauro-cit 
We can also remove some of the conditional logic we no longer need like in https://github.com/terraform-google-modules/terraform-example-foundation/blob/997d699b8b35f7539ac632230ef86033d630952a/3-networks-dual-svpc/modules/base_shared_vpc/variables.tf#L27

File: 3-networks-hub-and-spoke/modules/hierarchical_firewall_policy/main.tf [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  special = false
+}
+
+resource "google_compute_organization_security_policy" "policy" {
+  provider     = google-beta
+  display_name = "${var.name}-${random_string.suffix.result}"
+  parent       = var.parent
+}
+
+resource "google_compute_organization_security_policy_rule" "rule" {
```
[the remark this answers is not in the document]
Comment (mauro-cit) [reply 3-networks-hub-and-spoke/modules/hierarchical_firewall_policy/main.tf:33]: Issue created: https://github.com/terraform-google-modules/terraform-example-foundation/issues/746

[the remark this answers is not in the document]
Comment (mauro-cit) [reply 3-networks-hub-and-spoke/modules/hierarchical_firewall_policy/main.tf:33]: @bharathkkb is there any other module you want it to be upstreamed as well?

File: 3-networks-hub-and-spoke/modules/base_env/main.tf [hunk 6]
```diff
@@ -19,9 +19,8 @@ locals {
   restricted_project_number = data.google_project.restricted_host_project.number
   base_project_id           = data.google_projects.base_host_project.projects[0].project_id
   parent_id                 = var.parent_folder != "" ? "folders/${var.parent_folder}" : "organizations/${var.org_id}"
-  mode                      = var.enable_hub_and_spoke ? "spoke" : null
   bgp_asn_number            = var.enable_partner_interconnect ? "16550" : "64514"
-  enable_transitivity       = var.enable_hub_and_spoke && var.enable_hub_and_spoke_transitivity
+  enable_transitivity       = true
```
Comment (bharathkkb) [inline 3-networks-hub-and-spoke/modules/base_env/main.tf:23]: We should still have enable_hub_and_spoke_transitivity with default false.

File: 3-networks-hub-and-spoke/envs/shared/net-hubs.tf [hunk 7]
```diff
@@ -15,9 +15,9 @@
  */
 
 locals {
-  base_net_hub_project_id           = try(data.google_projects.base_net_hub[0].projects[0].project_id, null)
-  restricted_net_hub_project_id     = try(data.google_projects.restricted_net_hub[0].projects[0].project_id, null)
-  restricted_net_hub_project_number = try(data.google_projects.restricted_net_hub[0].projects[0].number, null)
+  base_net_hub_project_id           = try(data.google_projects.base_net_hub.projects[0].project_id, null)
+  restricted_net_hub_project_id     = try(data.google_projects.restricted_net_hub.projects[0].project_id, null)
+  restricted_net_hub_project_number = try(data.google_projects.restricted_net_hub.projects[0].number, null)
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply 3-networks-hub-and-spoke/envs/shared/net-hubs.tf:20]: @mauro-cit does this need a try? If so a comment will be useful for understanding usecase.

**Review by bharathkkb — APPROVED** [review]
LGTM with a suggestion

**Pull request merged**